pages/blog/_posts/2020-08-20-building-software-for-a-telecom-core-network.md

```diff
@@ -0,0 +1,157 @@
+
+```kotlin
+fun getVoicemail(user: User, uuid: String): Voicemail? {
+    val getVoicemailRequest = VoicemailProto.GetVoicemailRequest.newBuilder().setVoicemailId(uuid).build()
+    val voicemail = try {
+        blockingStub
+                .withOAuthTokenFor(user) // this function attaches the access token (from the consent screen)
+                .getVoicemail(getVoicemailRequest)
+    } catch (e: StatusRuntimeException) {
+        logger.warn("Error getting voicemail $uuid; ${e.message}")
+        throw e
+    }
+    if (voicemail.metadata) {
+        return Voicemail(
+            ... // we map the gRPC voicemail object to a VoiceBox specific object
+        )
+    } else {
+        logger.warn("No voicemail metadata for $uuid")
+        return null
+    }
+}
+```
+
+The first line of the function declares a `GetVoicemailRequest`, which is a class from the official Java client (this example is written in Kotlin, but Kotlin and Java are interoperable languages). Inside the `try`, we pass the OAuth credentials and the request to a `blockingStub`, which is of the type `VoicemailMediaServiceGrpc.VoicemailMediaServiceBlockingStub` (another class we get from the client). If the voicemail metadata is present, we have everything we need to return a VoiceBox specific `Voicemail` class, if not we have to log and return null. Unlike REST APIs, we don't have to think about writing client code and handling HTTP responses, this is taken care of by the client library.
+
+As we mentioned earlier, once we have the `Voicemail`, we send it to the subscriber using their own number as the sender:
+
+```kotlin
+fun sendMms(user: User, phone: Msisdn, audio: ByteString) = try {
+    val request = MmsProto.SendMessageFromSubscriberRequest
+            .newBuilder()
+            .addMessageContent(
+                MmsProto.MessageContent.newBuilder().setAudio(
+                    MmsProto.AudioContent.newBuilder().setWav(audio)
+                )
+            )
+            .setToE164(phone.toPhoneNumberProto()) // the receiver is the same as the sender
+            .setFromSubscriber(phone.toPhoneNumberProto()) // the sender is the same as the receiver
+            .build()
+    val response = blockingStub
+            .withOAuthTokenFor(user) // this function attaches the access token (from the consent screen)
+            .sendMessageFromSubscriber(request)
+    if (response.status == MmsProto.SendResponse.SendStatus.SEND_OK) {
+        true
+    } else {
+        logger.warn("Unable to send MMS to ${phone.e164}: ${response.description} (${response.status.name})")
+        false
+    }
+} catch (e: Exception) {
+    logger.warn("Unable to send MMS to ${phone.e164}: ${e.message}")
+    false
+}
+```
+
+This is also pretty straightforward, except for the part where the sender is also the receiver. In the future this will be a bit less confusing, since we’ll be sending the MMS from the product (“VoiceBox”) instead of the subscriber itself.
```

Comment:
*Personal opinions which might not reflect other people's opinions inc*

*Feedback 1*
I felt proud when I started reading the article and found the story being told interesting (you should probably fire me if I didn't).

*Feedback 2*
I read the article while I was out walking the dog. Somewhere around this bit I got a bit weary of reading the article and felt like taking a break from reading it. It might be because I'm already familiar with the content, but it felt a bit wordy to me, repeating a bit of what's written in code in plain english. This feels appropriate when writing a tutorial but felt verbose to me in an article geared at showcasing/telling a story about tech.

I'll try to highlight some of the sentences I found superfluous.

*Feedback 3*
Although not perfect I also think it would be great if we could a screenshot of a received mms with an audiofile or whatever, so you get the full story/feature. Probably at the beginning there with the first couple of pictures.

Reply:
Since I ported my number to Wotel, I can no longer use this feature, so I can't take a screenshot 😅